sunscreen_zkp_backend/src/bulletproofs.rs

```diff
@@ -145,10 +145,26 @@ impl BulletproofsCircuit {
         CS: ConstraintSystem,
         I: Fn(usize) -> Option<Scalar>,
     {
+        let mut unprocessed_child_count = graph.node_indices().map(|n| {
+            graph.neighbors(n).count()
+        }).collect::<Vec<usize>>();
+        
         // The graph won't actually be mutated.
         forward_traverse(&graph.0, |query, idx| {
             let node = query.get_node(idx).unwrap();
 
+            // Each linear combination object in Bulletproofs has a Vec
+            // in it and thus ain't cheap to store. As such, we reference
+            // count the output of a given node when all its children have
+            // been processed.
+            let ref_count = |nodes: &mut Vec<Option<Node>>, idx: NodeIndex, unprocessed_child_count: &mut Vec<usize>| {
+                unprocessed_child_count[idx.index()] -= 1;
```

Comment:
Is it logically impossible for any of these entries `unprocessed_child_count[idx.index()]` to start at `0`?

Reply:
If not, I don't think the logic really changes, just calling it out that we'd want a saturating sub, since these are `usize`s

Reply:
The only nodes with an initialized zero count are those with no children. However, this code will never get called in such cases, because there are no children to call it 😄